### PR TITLE
Update DatabaseConnection.java

### PR DESCRIPTION
**Pull request by Cdele**

### CODE REVIEW

File: backend/src/main/java/pw/wp6/avocado_toast/invoker/DatabaseConnection.java [hunk 1]
```diff
@@ -22,28 +22,46 @@
...
+                        "  name                 TEXT     NOT NULL,\n" +
+                        "  balance              INTEGER  NOT NULL,\n" +
+                        "  overdraft            BOOLEAN  NOT NULL,\n" +
+                        "  ssn                  INTEGER  NOT NULL,\n" +
+                        "  account_type         TEXT     NOT NULL\n" +
                         ");");
                 stmt.executeUpdate(
-                        "INSERT OR IGNORE INTO users (id, name, username, password, ssn, account_type)\n" +
-                                "VALUES (0, 'Admin', 'admin', '" +
-                                new BCryptPasswordEncoder().encode("admin") +
```
Comment (flaviut) [inline backend/src/main/java/pw/wp6/avocado_toast/invoker/DatabaseConnection.java]: Don't get rid of this line. If you do, that breaks logging in as the admin user.

File: backend/src/main/java/pw/wp6/avocado_toast/invoker/DatabaseConnection.java [hunk 2]
```diff
@@ -22,28 +22,46 @@
...
+                        "  analyst_username     TEXT     UNIQUE NOT NULL,\n" +
+                        "  analyst_password     TEXT     NOT NULL,\n" +
+                        "  name                 TEXT     NOT NULL,\n" +
+                        "  account_type         TEXT     NOT NULL\n" +
+                        ");");
+                 stmt.execute("CREATE TABLE IF NOT EXISTS Customer\n" +
+                        "(\n" +
+                        "  id                   INTEGER  PRIMARY KEY,\n" +
+                        "  customer_user_id     INTEGER  NOT NULL REFERENCES Transactions (id),\n" +
+                        "  Customer_username    TEXT     UNIQUE NOT NULL,\n" +
```
Comment (flaviut) [inline backend/src/main/java/pw/wp6/avocado_toast/invoker/DatabaseConnection.java]: Capitalize consistently.

File: backend/src/main/java/pw/wp6/avocado_toast/invoker/DatabaseConnection.java [hunk 3]
```diff
@@ -22,28 +22,46 @@
...
+                stmt.execute("CREATE TABLE IF NOT EXISTS Banker\n" +
                         "(\n" +
-                        "  id        INTEGER PRIMARY KEY,\n" +
-                        "  user_id   INTEGER  NOT NULL REFERENCES users (id),\n" +
-                        "  merchant  TEXT     NOT NULL,\n" +
-                        "  amount    DECIMAL  NOT NULL,\n" +
-                        "  date_time DATETIME NOT NULL\n" +
+                        "  id                   INTEGER  PRIMARY KEY,\n" +
+                        "  banker_user_id       INTEGER  NOT NULL REFERENCES Transactions (id),\n" +
+                        "  banker_username      TEXT     UNIQUE NOT NULL,\n" +
```
Comment (flaviut) [inline backend/src/main/java/pw/wp6/avocado_toast/invoker/DatabaseConnection.java]: Get rid of all the "banker_". Same for the other tables.

This prefix makes it harder to write queries.

File: backend/src/main/java/pw/wp6/avocado_toast/invoker/DatabaseConnection.java [hunk 4]
```diff
@@ -22,28 +22,46 @@
...
+                        "  analyst_user_id      INTEGER  NOT NULL REFERENCES Transactions (id),\n" +
+                        "  analyst_username     TEXT     UNIQUE NOT NULL,\n" +
+                        "  analyst_password     TEXT     NOT NULL,\n" +
+                        "  name                 TEXT     NOT NULL,\n" +
+                        "  account_type         TEXT     NOT NULL\n" +
+                        ");");
+                 stmt.execute("CREATE TABLE IF NOT EXISTS Customer\n" +
+                        "(\n" +
+                        "  id                   INTEGER  PRIMARY KEY,\n" +
+                        "  customer_user_id     INTEGER  NOT NULL REFERENCES Transactions (id),\n" +
```
Comment (flaviut) [inline backend/src/main/java/pw/wp6/avocado_toast/invoker/DatabaseConnection.java]: This foreign key makes no sense. Each user has exactly one transaction?

Rework this please, to have a 1-n customer-transaction relation.

File: backend/src/main/java/pw/wp6/avocado_toast/invoker/DatabaseConnection.java [hunk 5]
```diff
@@ -22,28 +22,46 @@
...
+                        "  Customer_username    TEXT     UNIQUE NOT NULL,\n" +
+                        "  customer_password    TEXT     NOT NULL,\n" +
+                        "  name                 TEXT     NOT NULL,\n" +
+                        "  balance              INTEGER  NOT NULL,\n" +
```
Comment (flaviut) [inline backend/src/main/java/pw/wp6/avocado_toast/invoker/DatabaseConnection.java]: Denormalizing this doesn't serve any propose. Remove the balance and overdraft columns.

The balance is calculated on-demand with a SELECT & SUM() query.

File: backend/src/main/java/pw/wp6/avocado_toast/invoker/DatabaseConnection.java [hunk 6]
```diff
@@ -22,28 +22,46 @@
...
+                        "  account_type         TEXT     NOT NULL\n" +
+                        ");");
+                stmt.execute("CREATE TABLE IF NOT EXISTS Analyst\n" +
+                        "(\n" +
+                        "  id                   INTEGER  PRIMARY KEY,\n" +
+                        "  analyst_user_id      INTEGER  NOT NULL REFERENCES Transactions (id),\n" +
+                        "  analyst_username     TEXT     UNIQUE NOT NULL,\n" +
+                        "  analyst_password     TEXT     NOT NULL,\n" +
+                        "  name                 TEXT     NOT NULL,\n" +
+                        "  account_type         TEXT     NOT NULL\n" +
```
Comment (flaviut) [inline backend/src/main/java/pw/wp6/avocado_toast/invoker/DatabaseConnection.java]: The account type column is redundant. There can only be customers in the customer table.

**Comment by flaviut** [comment]
This PR denormalizes the database. If we're not required to have a normalized database that's fine, but if we are, this won't work.

Read the requirements and see if a normalized database is required.

**Comment by Cdele** [comment]
@flaviut check this when you can pls

File: backend/src/main/java/pw/wp6/avocado_toast/invoker/DatabaseConnection.java [hunk 7]
```diff
@@ -21,29 +21,47 @@
...
+                        "  amount                        INTEGER     NOT NULL,\n" +
+                        "  date_time                     DATETIME    NOT NULL\n" +
                         ");");
                 stmt.executeUpdate(
-                        "INSERT OR IGNORE INTO users (id, name, username, password, ssn, account_type)\n" +
-                                "VALUES (0, 'Admin', 'admin', '" +
-                                new BCryptPasswordEncoder().encode("admin") +
-                                "', '000-00-0000', 'BANKER');");
+                        "INSERT OR IGNORE INTO Banker (id, username, password, name, account_type)\n" +
+                                "VALUES (0, 'Admin',
```
Comment (flaviut) [inline backend/src/main/java/pw/wp6/avocado_toast/invoker/DatabaseConnection.java]: Syntax error on this line

File: backend/src/main/java/pw/wp6/avocado_toast/invoker/DatabaseConnection.java [hunk 8]
```diff
@@ -21,29 +21,47 @@
...
+                        "  password                      TEXT     NOT NULL,\n" +
+                        "  name                          TEXT     NOT NULL,\n" +
+                        "  ssn                           INTEGER  NOT NULL,\n" +
+                        "  account_type                  TEXT     NOT NULL\n" +
+                        ");");
+                stmt.execute("CREATE TABLE IF NOT EXISTS Transactions\n" +
+                        "(\n" +
+                        "  id                            INTEGER     PRIMARY KEY,\n" +
+                        "  customer_user_id              INTEGER     NOT NULL REFERENCES Customer (id),\n" +
+                        "  banker_user_id                INTEGER     NOT NULL REFERENCES Banker (id),\n" +
```
Comment (flaviut) [inline backend/src/main/java/pw/wp6/avocado_toast/invoker/DatabaseConnection.java]: Bankers and don't have transactions in their accounts

File: backend/src/main/java/pw/wp6/avocado_toast/invoker/DatabaseConnection.java [hunk 9]
```diff
@@ -21,29 +21,47 @@
...
+                        "  analyst_user_id               INTEGER     NOT NULL REFERENCES Analyst (id),\n" +
+                        "  merchant                      TEXT        NOT NULL,\n" +
+                        "  amount                        INTEGER     NOT NULL,\n" +
+                        "  date_time                     DATETIME    NOT NULL\n" +
```
Comment (flaviut) [inline backend/src/main/java/pw/wp6/avocado_toast/invoker/DatabaseConnection.java]: I don't know if sqlite supports date time.. You might have to use strings

**Review by flaviut — CHANGES_REQUESTED** [review]

**Review by flaviut — APPROVED** [review]

**Pull request merged**